modules/core/src/test/java/org/locationtech/jts/algorithm/distance/MatrixStorageImplTests.java

```diff
@@ -0,0 +1,288 @@
+  {
+    MatrixStorage mat = new CsrMatrix(4, 6, 0d, 8);
+    runOrderedTest(mat);
+    mat = new CsrMatrix(4, 6, 0d, 8);
+    runUnorderedTest(mat);
+  }
+  @Test
+  public void testHashMapMatrix()
+  {
+    MatrixStorage mat = new HashMapMatrix(4, 6, 0d);
+    runOrderedTest(mat);
+    mat = new HashMapMatrix(4, 6, 0d);
+    runUnorderedTest(mat);
+  }
+  @Test
+  public void testRectMatrix()
+  {
+    MatrixStorage mat = new RectMatrix(4, 6, 0d);
+    runOrderedTest(mat);
+    mat = new RectMatrix(4, 6, 0d);
+    runUnorderedTest(mat);
+  }
+
+  private static void runOrderedTest(MatrixStorage mat) {
+    mat.set(0, 0, 10);
+    mat.set(0, 1, 20);
+    mat.set(1, 1, 30);
+    mat.set(1, 3, 40);
+    mat.set(2, 2, 50);
+    mat.set(2, 3, 60);
+    mat.set(2, 4, 70);
+    mat.set(3, 5, 80);
+
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 0, 0, 10d, mat.get(0, 0)), 10d, mat.get(0, 0));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 0, 1, 20d, mat.get(0, 1)), 20d, mat.get(0, 1));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 1, 1, 30d, mat.get(1, 1)), 30d, mat.get(1, 1));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 1, 3, 40d, mat.get(1, 3)), 40d, mat.get(1, 3));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 2, 2, 50d, mat.get(2, 2)), 50d, mat.get(2, 2));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 2, 3, 60d, mat.get(2, 3)), 60d, mat.get(2, 3));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 2, 4, 70d, mat.get(2, 4)), 70d, mat.get(2, 4));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 3, 5, 80d, mat.get(3, 5)), 80d, mat.get(3, 5));
+
+  }
+
+  private static void runUnorderedTest(MatrixStorage mat) {
+    mat.set(0, 0, 10);
+    mat.set(3, 5, 80);
+    mat.set(0, 1, 20);
+    mat.set(2, 4, 70);
+    mat.set(1, 1, 30);
+    mat.set(2, 3, 60);
+    mat.set(2, 2, 50);
+    mat.set(1, 3, 40);
+
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 0, 0, 10d, mat.get(0, 0)), 10d, mat.get(0, 0));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 0, 1, 20d, mat.get(0, 1)), 20d, mat.get(0, 1));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 1, 1, 30d, mat.get(1, 1)), 30d, mat.get(1, 1));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 1, 3, 40d, mat.get(1, 3)), 40d, mat.get(1, 3));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 2, 2, 50d, mat.get(2, 2)), 50d, mat.get(2, 2));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 2, 3, 60d, mat.get(2, 3)), 60d, mat.get(2, 3));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 2, 4, 70d, mat.get(2, 4)), 70d, mat.get(2, 4));
+    assertEquals(String.format("%1$d -> %2$d = %4$f /= %3$f", 3, 5, 80d, mat.get(3, 5)), 80d, mat.get(3, 5));
+
+  }
+
+  /**
+   * Abstract base class for storing 2d matrix data
+   */
+  private abstract static class MatrixStorage {
```

Comment:
It's unusual to duplicate implementation code in unit tests, since this doesn't actually test the code in the library?

Reply:
I did not see another way to test and verify that `MatrixStorage` implementations work correctly while keeping it private in jts-core.

Reply:
Yeah, that's a bit of a limitation in the Java package model.  I usually make things like this package-private, so they can be unit tested.